packages/react-query/src/utils/structuralSharing.ts

```diff
@@ -0,0 +1,154 @@
+/**
+ * Inspired by @tanstack/query-core utils.ts
+ * Modified to support Date, Map, Set and NaN comparisons
+ *
+ * This function returns `a` if `b` is deeply equal.
+ * If not, it will replace any deeply equal children of `b` with those of `a`.
+ */
+export function replaceEqualDeep(a: any, b: any): any {
```

Comment:
I think we should allow custom comparators here too - for instance if you use any custom data types like `Temporal.PlainDate` etc?

Could be a factory function like


Also, I'm not sold on the naming 🤔 🤔

Reply:
I changed to a factory function + a default comparator, let me know what you think.

I  also renamed but I'm still not sure it's great.